homeassistant/components/slack/notify.py

```diff
@@ -112,6 +164,35 @@ async def _async_send_local_file_message(self, path, targets, message, title):
         except SlackApiError as err:
             _LOGGER.error("Error while uploading file-based message: %s", err)
 
+    async def _async_send_remote_file_message(
+        self, url, targets, message, title, *, username=None, password=None
+    ):
+        """Upload a remote file (with message) to Slack."""
+        if not self._hass.config.is_allowed_external_url(url):
+            _LOGGER.error("URL is not allowed: %s", url)
+            return
+
+        try:
+            file_as_bytes = await _async_get_remote_file_contents(
+                self._hass, url, username=username, password=password
+            )
+        except ClientError as err:
+            _LOGGER.error("Error while retrieving %s: %s", url, err)
+            return
+
+        filename = _async_get_filename_from_url(url)
+
+        try:
+            await self._client.files_upload(
```

Comment:
Does this method accept a stream for the upload too? In that case a lot better to just pass the stream from the source.

Reply:
It doesn't appear so: https://api.slack.com/methods/files.upload

Reply:
The API supports streaming as that's how HTTP commands arrive at any server :) 

The question is if the API exposes a feature for it. 

For example, if the API is using aiohttp… https://docs.aiohttp.org/en/stable/client_quickstart.html#streaming-uploads

Reply:
That’s what I mean: there’s no indication of a stream-related parameters in this API call. It uses `aiohttp` under the hood, but I have no way of getting at that.

Reply:
You need to look at their Python API, not their HTTP API.

Reply:
Let's make sure we're on the same page – I pasted a link to Slack's HTTP API, but when I say

> there’s no indication of stream-related parameters in this API call

...I'm speaking about [`python-slackclient`](https://github.com/slackapi/python-slackclient). The `files_upload` method  used takes a `file` argument that can either be a string (tests show that this only works with local file paths) or an `IOBytes` (https://github.com/slackapi/python-slackclient/blob/2f657eac351f63d53720868b5285ee11edc4f6cb/slack/web/client.py#L1192-L1194) – I don't see any indication that I can pass a parameter related to a streaming upload.

Reply:
What about this? https://stackoverflow.com/a/60721587

Reply:
That is a Node.js package – it has a mechanism for reading a file from disk as a stream, but that has nothing to do with `python-slackclient`.

I'm still trying to understand what the problem is here. Are you concerned that the remote file upload mechanism in this PR downloads an entire remote file into memory (as bytes) before using `python-slackclient` to upload it? If so, I can investigate whether I can access some lower-level API coroutine in `python-slackclient`...

Reply:
Oh right that was the node one.

The problem is two-fold:
 - We don't want to store the whole file in memory, as it can cause our instance to run out of memory and Home Assistant to be killed. Not every download will also include the size of the file that we're downloading.
 - It's a lot slower if we have to first wait to download a file and then start uploading it. It's a lot faster to be able to start the upload the moment we get the first byte in.

Reply:
Got it. Thanks for the clarification. I’ll see what can be done.

Reply:
@balloob I've done a ton of digging and can't find anything in `python-slackclient` that would allow me to do a streaming upload from a remote URL. Not sure where to go from here. 🤷

Reply:
I think that this will work:

You will need to call the [files.upload](https://api.slack.com/methods/files.upload) url directly with aiohttp (bypassing their client) and then use aiohttp's [`FormData`](https://docs.aiohttp.org/en/stable/client_reference.html#aiohttp.FormData). 

Could be something like this (untested)

```python
from aiohttp import FormData

resp = await session.get('http://python.org')

data = FormData({
  "token": "abcd",
  "channels": "qwer",
  "title": "asdf",
  "initial_comment": "zxcczxc",
})
data.add_field("file", resp.content)

await session.post('https://slack.com/api/files.upload', data=data)
```

Reply:
Thanks for the help, @balloob – I'll investigate.

Reply:
Initial tests aren't working – Slack keeps returning a `{"ok":false,"error":"no_file_data"}` response.

Going to move this to draft state, as I don't have time to work on it now. Will pick back up later.

Reply:
Whew, got it!